### PR TITLE
GIX-1212: Refactor ProposalsFilterModal.svelte

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

Refactor the ProposalsFilterModal.svelte so that it can be reused in SNS.

# Changes

Main change:
* New component FilterModal.svelte. This is the UI only part of ProposalsFilterModal component.
* Refactor ProposalsFilterModal to use FilterModal
* Rename ProposalsFilterModal to NnsProposalsFilterModal

Other minor changes:
* New folder `modals/common` where I moved ConfirmationModal and FilterModal.

# Tests

* New test for new FilterModal.svelte
* Fix broken tests after refactor and file movements.

### CODE REVIEW

**Comment by lmuntaner** [comment]
@peterpeterparker @mstrasinskis please review

No UI nor UX changes

File: frontend/src/lib/modals/common/FilterModal.svelte [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  {#if filters}
+    <div class="filters">
+      {#each filters as { id, name, checked } (id)}
+        <Checkbox inputId={id} {checked} on:nnsChange={() => onChange(id)}
+          >{name}</Checkbox
+        >
+      {/each}
+    </div>
+  {:else}
+    <Spinner />
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/common/FilterModal.svelte:39]: Actually that's a UI change 😉.

UI wise the pattern that displays the spinner and then the modal is visually more pleasant. e.g. `SnsIncreaseStakeNeuronModal`.

Can you do that here or in another PR?

File: frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte [hunk 2]
```diff
@@ -17,15 +15,38 @@
...
+    value: Topic | ProposalRewardStatus | ProposalStatus
+  ) => Filter<Topic | ProposalRewardStatus | ProposalStatus>;
+  $: mapToFilter = (value: Topic | ProposalRewardStatus | ProposalStatus) => {
+    return {
+      id: String(value),
+      value,
+      name:
+        keyOfOptional({
+          obj: i18nKeys,
+          key: filters?.[value] ?? "Unspecified",
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte:44]: "Unspecified" as `string` is correct?

If correct, i18n key is missing.

File: frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte [hunk 3]
```diff
@@ -72,8 +94,13 @@
     }
   };
 
-  const onChange = (filter: Topic | ProposalRewardStatus | ProposalStatus) =>
-    applyFilterChange(filter);
+  const onChange = ({
+    detail: { filter },
+  }: CustomEvent<{
+    filter: Filter<Topic | ProposalRewardStatus | ProposalStatus>;
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte:100]: `filter` can be `undefined` here

File: frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte [hunk 4]
```diff
@@ -17,15 +15,38 @@
...
+  $: i18nKeys =
+    props?.category !== undefined
+      ? keyOf({ obj: $i18n, key: props?.category })
+      : undefined;
   let filters: ProposalsFilters | undefined;
-  let filtersValues: number[];
+  let filtersValues: Filter<Topic | ProposalRewardStatus | ProposalStatus>[];
   let selectedFilters: (Topic | ProposalRewardStatus | ProposalStatus)[];
 
+  let mapToFilter: (
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte:34]: this needs to be a reactive variable? a `const` initialized on mount would not do?

File: frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte [hunk 5]
```diff
@@ -17,15 +15,38 @@
...
   import { keyOf, keyOfOptional } from "$lib/utils/utils";
   import { DEPRECATED_TOPICS } from "$lib/constants/proposals.constants";
+  import FilterModal from "../common/FilterModal.svelte";
+  import type { Filter } from "$lib/types/filters";
 
   export let props: ProposalsFilterModalProps | undefined;
 
   let visible: boolean;
   let category: string;
+  let i18nKeys: unknown;
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte:25]: - it can be `undefined` too
- `unknown` for simplicity reason - i.e. to avoid the hustle of these enums or is it a why to type it?

File: frontend/src/lib/modals/neurons/ConfirmRemoveCurrentUserHotkey.svelte [hunk 6]
```diff
@@ -1,6 +1,6 @@
 <script lang="ts">
   import { i18n } from "$lib/stores/i18n";
-  import ConfirmationModal from "../ConfirmationModal.svelte";
+  import ConfirmationModal from "../common/ConfirmationModal.svelte";
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/neurons/ConfirmRemoveCurrentUserHotkey.svelte:3]: nth: `$lib/modals/common/ConfirmationModal.svelte`

File: frontend/src/lib/modals/common/FilterModal.svelte (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/lib/modals/common/FilterModal.svelte:39]: Haha, true! Sorry.

I see what you mean. Sure sounds good!

File: frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte (hunk 2 above)
Comment (lmuntaner) [reply frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte:44]: The key exists, but I think that it's better to default to `""` and show the `Unspecified` if `keyOfOptional` returns undefined.

File: frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte (hunk 3 above)
Comment (lmuntaner) [reply frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte:100]: Can it? Do you mean for type safety? Or that Svelte might set it to `undefined`?

File: frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte (hunk 4 above)
Comment (lmuntaner) [reply frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte:34]: Yes, it needs to be. At the moment it's true that in NNS all the filters are hard coded as enums. But in SNS, the filter of the Topics comes from the backend and are loaded. I wanted the NNS to do be able to handle that also.

It's not necessary now, but I believed that it doesn't hurt and it prepares us for future changes.

File: frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte (hunk 5 above)
Comment (lmuntaner) [reply frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte:25]: Correct, `unknown` for simplicity. Polimorphism in Typescript is not very nice...

Comment (lmuntaner) [reply frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte]: Added `undefined`

File: frontend/src/lib/modals/neurons/ConfirmRemoveCurrentUserHotkey.svelte (hunk 6 above)
Comment (lmuntaner) [reply frontend/src/lib/modals/neurons/ConfirmRemoveCurrentUserHotkey.svelte]: done

**Comment by lmuntaner** [comment]
@peterpeterparker ready for another review

File: frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte (hunk 3 above)
Comment (peterpeterparker) [reply frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte:100]: I meant this because the `dispatch` can emit `undefined` or I am missing something?

In `FilterModal.svelte`:

```
const onChange = (id: string) => {
    const filter = filters?.find((f) => f.id === id);
    dispatch("nnsChange", { filter });
  };
```

File: frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte (hunk 4 above)
Comment (peterpeterparker) [reply frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte:34]: Gotcha, thanks for the explanation

File: frontend/src/lib/modals/common/FilterModal.svelte [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  // Source: https://github.com/dummdidumm/rfcs/blob/ts-typedefs-within-svelte-components/text/ts-typing-props-slots-events.md#solution
+  type T = $$Generic;
+  // `undefined` means the filters are not loaded yet.
+  export let filters: Filter<T>[] | undefined;
+  export let visible = true;
+
+  let loading: boolean;
+  $: loading = filters === undefined;
+
+  onMount(async () => {
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/common/FilterModal.svelte:20]: you don't need the `async` here I think

File: frontend/src/lib/modals/common/FilterModal.svelte [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    if (filters !== undefined) {
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/common/FilterModal.svelte:21]: nth: `nonNullish`

File: frontend/src/lib/modals/common/FilterModal.svelte [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    const filter = filters?.find((f) => f.id === id);
+    dispatch("nnsChange", { filter });
+  };
+
+  const filter = () => {
+    dispatch("nnsConfirm");
+  };
+</script>
+
+{#if nonNullish(filters)}
```
Comment (peterpeterparker) [inline frontend/src/lib/modals/common/FilterModal.svelte:50]: should you check `loading` here?

**Review by peterpeterparker — APPROVED** [review]
Answered one and added few minor comments but already approving so you can move one.
Thanks!

File: frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte (hunk 3 above)
Comment (lmuntaner) [reply frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte:100]: Ah, no, you're right. To be safe it shuold be that. Thanks!

I believe we can connect those two types, I think I read it somewhere...

Comment (lmuntaner) [reply frontend/src/lib/modals/proposals/NnsProposalsFilterModal.svelte]: done

File: frontend/src/lib/modals/common/FilterModal.svelte (hunk 9 above)
Comment (lmuntaner) [reply frontend/src/lib/modals/common/FilterModal.svelte]: yes, thanks!

File: frontend/src/lib/modals/common/FilterModal.svelte (hunk 8 above)
Comment (lmuntaner) [reply frontend/src/lib/modals/common/FilterModal.svelte:21]: done
